rel/graph.c

```diff
@@ -0,0 +1,217 @@
+  this->Gfx_list01 = (Gfx*)GET_DYNAMIC_OFS(OVERLAY_OFS);
+  this->Gfx_list04 = (Gfx*)GET_DYNAMIC_OFS(LINE_OPA_OFS);
+  this->Gfx_list05 = (Gfx*)GET_DYNAMIC_OFS(WORK_OFS);
+  this->Gfx_list07 = (Gfx*)GET_DYNAMIC_OFS(POLY_OPA_OFS);
+  this->Gfx_list08 = (Gfx*)GET_DYNAMIC_OFS(POLY_XLU_OFS);
+  this->Gfx_list09 = (Gfx*)GET_DYNAMIC_OFS(FONT_OFS);
+
+  this->gfxsave = NULL;
+}
+
+/* TODO: This requires multiple GAME classes to be implemented before it will be matching */
+static DLFTBL_GAME* game_get_next_game_dlftbl(GAME* game) {
+  return NULL;
+}
+
+extern void graph_ct(GRAPH* this) {
+  bzero(this, sizeof(GRAPH));
+  this->frame_counter = 0;
+  this->cfb_bank = 0;
+  zurumode_init();
+  SETREG(SREG, 33, GETREG(SREG, 33) & ~2);
+  SETREG(SREG, 33, GETREG(SREG, 33) & ~1);
+  GRAPH_SET_DOING_POINT(this, CT);
+}
+
+extern void graph_dt(GRAPH* this) {
+  GRAPH_SET_DOING_POINT(this, DT);
+  zurumode_cleanup();
+}
+
+static void graph_task_set00(GRAPH* this) {
+  ucode_info ucode[2];
+
+  GRAPH_SET_DOING_POINT(this, WAIT_TASK);
+  GRAPH_SET_DOING_POINT(this, WAIT_TASK_FINISHED);
+  if (ResetStatus < IRQ_RESET_DELAY) {
+    ucode[0].type = UCODE_TYPE_POLY_TEXT;
+    ucode[1].type = UCODE_TYPE_SPRITE_TEXT;
+    ucode[0].ucode_p = ucode_GetPolyTextStart();
+    ucode[1].ucode_p = ucode_GetSpriteTextStart();
+    //JW_BeginFrame();
+    //emu64_init();
+    //emu64_set_ucode_info(2, ucode);
+    //emu64_set_first_ucode(ucode[0].ucode_p);
+    // emu64_taskstart(this->Gfx_list05); /* work data */
+    // emu64_cleanup();
+    // JW_EndFrame();
+    frame++;
+  }
+}
+
+static int graph_draw_finish(GRAPH* this) {
+  int err = 0;
+  OPEN_DISP(this);
+
+  gSPBranchList(NOW_WORK_DISP++, this->Gfx_list10);
+  gSPBranchList(NOW_SHADOW_DISP++, this->Gfx_list08);
+  gSPBranchList(NOW_POLY_XLU_DISP++, this->Gfx_list11);
+  gSPBranchList(NOW_LIGHT_DISP++, this->Gfx_list00);
+  gSPBranchList(NOW_LINE_OPA_DISP++, this->Gfx_list01);
+  gSPBranchList(NOW_OVERLAY_DISP++, this->Gfx_list09);
+  gSPBranchList(NOW_FONT_DISP++, this->Gfx_list07);
+  gSPBranchList(NOW_POLY_OPA_DISP++, this->Gfx_list04);
+  gDPPipeSync(NOW_LINE_OPA_DISP++);
+  gDPFullSync(NOW_LINE_OPA_DISP++);
+  gSPEndDisplayList(NOW_LINE_OPA_DISP++);
+
+  CLOSE_DISP(this);
+
+  if (*(u16*)GET_DYNAMIC_OFS(SYSDYNAMIC_HEADER_OFS) != SYSDYNAMIC_START_MAGIC) {
+    _dbg_hungup(__FILE__, 417);
+  }
+
+  if (*(u16*)GET_DYNAMIC_OFS(SYSDYNAMIC_EPILOG_OFS) != SYSDYNAMIC_END_MAGIC) {
+    _dbg_hungup(__FILE__, 425);
+    err = 1;
+  }
+
+  err = THA_GA_isCrash(&this->line_translucent_thaga) != 0;
+  err = THA_GA_isCrash(&this->overlay_thaga) != 0;
+  err = THA_GA_isCrash(&this->line_opaque_thaga) != 0;
+  err = THA_GA_isCrash(&this->polygon_opaque_thaga) != 0;
+  err = THA_GA_isCrash(&this->polygon_translucent_thaga) != 0;
+  err = THA_GA_isCrash(&this->font_thaga) != 0;
+  err = THA_GA_isCrash(&this->shadow_thaga) != 0;
+  err = THA_GA_isCrash(&this->light_thaga) != 0;
+
+  return err;
```

Comment:
I kind of hope this doesn't match since at the moment it only returns `true` if the light arena has crashed!

Reply:
I think it's a check if any of the arenas are in an undefined state. So the return type is `BOOL isError` or something along those lines.